src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx

```diff
@@ -1288,10 +1295,9 @@ namespace xclcpuemhal2
       if (mIsKdsSwEmu && mSWSch && mCore)
       {
         mSWSch->fini_scheduler_thread();
-        delete mCore;
-        mCore = nullptr;
-        delete mSWSch;
-        mSWSch = nullptr;
+        mCore.reset();
+        mSWSch.reset();
```

Comment:
do we really require reset in every reset?

Reply:
well, Shim class shall never have multiple mcore, mSwsh in it, the destructor responsibility is shared accross resetProgram() & shim dtor. So the same discipline is maintained without disturbing.

Reply:
This is the kind of knowledge you should add in some comments to explain the code more.
Since you have spent a lot of time on this it would be nice to have everybody benefit of this knowledge.
So, feel free to add a lot of comments everywhere. :-)